### ISSUE TITLE
Supports for hashtable iteration: functions that retrieve linked lists of rooms, players, & items.

### ISSUE THREAD

**Issue by liuyifan-eric** [uchicago-cs/chiventure#214]
@jonahdf reported on slack that they (checkpointing) are getting a global variable redeclaration error when calling our (game-state) ITER_ALL_PLAYERS and ITER_ALL_ROOMS macros. 
We need to update the macros to solve this problem, so that the ITER macros can be used multiple times in each function.

**Comment by mpcsh** [comment]
@liuyifan-eric We investigated this during team meeting and determined that patching the macro will either be impossible or a massive pain. The proper solution here is to create a functions that return lists of whatever object is desired (e.g. `get_all_players()`, `get_all_rooms()`, etc).

**Comment by borjasotomayor** [comment]
As a stopgap measure, you could change the macro to something like this:

```
ITER_ALL_ROOMS(game, curr_room, room)
```

And require that the user of `ITER_ALL_ROOMS` declare a room_t* variable to use as an iterator. However, this increases the cognitive load of the user of `ITER_ALL_ROOMS` and, in the long run, what we need are functions that return lists (or simply arrays of rooms/items/etc., which would be easier to build and operate on than linked lists)

**Comment by liuyifan-eric** [comment]
As a stopgap measure, I updated the macros so that any code with macros does not break. (It's still ITER_ALL_ROOMS(game, curr_room), but uses a global variable for stopgap.)

**Comment by borjasotomayor** [comment]
Global variables are strictly forbidden (see https://uchicago-cs.github.io/cmsc23300/style_guide.html#global-variables, linked from the developer guidelines at https://github.com/uchicago-cs/chiventure/blob/master/docs/developer.md). 

Please revert that commit, and do not implement any stopgap measure that relies on global variables.

**Comment by liuyifan-eric** [comment]
> Global variables are strictly forbidden (see https://uchicago-cs.github.io/cmsc23300/style_guide.html#global-variables, linked from the developer guidelines at https://github.com/uchicago-cs/chiventure/blob/master/docs/developer.md).
> 
> Please revert that commit, and do not implement any stopgap measure that relies on global variables.

Sorry about that. 
This is on the feature branch and not pushed to the dev branch, so I will just abandon the branch.

**Comment by liuyifan-eric** [comment]
All plans on the macros are terminated. Current macros in game-state/dev and master will not be modified for now in order to prevent anything from breaking. After discussing with the checkpointing team, we will turn to functions that retrieve a linked list of rooms/players/items.

**Comment by liuyifan-eric** [comment]
Updated room related iteration functions:
`/*
 * Function to get a linked list (utlist) of all the rooms in the game
 *
 * Parameters:
 *  game
 *
 * Returns:
 *  linked list of pointers to rooms (the head element)
 */
room_list_t *get_all_rooms(game_t *game);

/*
 * Function to delete a linked list (utlist) retrieved from get_all_rooms()
 *
 * Parameters:
 *  linked list of pointers to rooms
 *
 * Returns:
 *  SUCCESS on success, FAILURE if an error occurs.
 */
int delete_room_llist(room_list_t *list);`
Tests passed
We will move on to implement the same functions on other structs

**Comment by liuyifan-eric** [comment]
Added attributes functions

**Comment by liuyifan-eric** [comment]
Added item functions;
need to add a function for all items in the game, as required by checkpointing.

**Comment by liuyifan-eric** [comment]
Added a function that retrieves all the items in the game.

**Comment by liuyifan-eric** [comment]
All above are done on a hotfix branch of hotfix/iter_func, and pull request is here: #227

**Comment by liuyifan-eric** [comment]
Closing comment:
We added functions that iterate game-state structs and return a list of the contents in the hashtable, including:
- `get_all_items()`
- `get_all_items_in_room()`
- `get_all_items_in_inventory()`
- `get_all_rooms()`
- `get_all_attributes_in_item()`
These functions replaced the old ITER macros.
The abandoned macros are still left in the code to avoid any unexpected breaks. They will be removed after everyone updates the macros to the functions.
